README
-----Ruby/SerialPort-----

-- Description --

Ruby/SerialPort is a Ruby library that provides a class for using
RS-232 serial ports.  This class also contains low-level functions to
check and set the current state of the signals on the line.

The native Windows version of this library supports Microsoft's Visual C++, Borland's C++, and MinGW compilers.

-- Installation --

This gem is hosted at Gemcutter. Before you can install this, or any other Gemcutter gem you must
add it as a gem source:
  gem sources -a http://gemcutter.org

Then you can install the gem as normal:
  sudo gem install serialport

-- Testing --

* test/miniterm.rb

Ruby's copy of miniterm.c !


-- API --

  **** Class SerialPort, Parent IO ****

    ** Class constants **

       VERSION -> aString (this release is "0.6")
       NONE, HARD, SOFT, SPACE, MARK, EVEN, ODD -> anInteger

    ** Class methods **

      * new(port_num [, modem_parameters]) -> aSerialPort
      * open(port_num [, modem_parameters]) -> aSerialPort
      * open(port_num [, modem_parameters]) {|aSerialPort| block} ->
                value of the block

        port_num -> anInteger: port number, 0 for first port which is
                      "/dev/ttyS0" on GNU/Linux and "COM1" on Windows,
                or  aString: file name of the device (example: "/dev/ttyS2")

        Optional modem_parameters:

        baudrate -> anInteger: from 50 to 256000, depends on platform.

        databits -> anInteger: from 5 to 8 (4 is allowed on Windows)

        stopbits -> anInteger: 1 or 2 (1.5 is not supported)

        parity -> anInteger: SerialPort::NONE, SerialPort::EVEN,
                  SerialPort::ODD, SerialPort::MARK, SerialPort::SPACE
                  (MARK and SPACE are not supported on Posix)

        Raise an argError on bad argument.

        SerialPort::new and SerialPort::open without a block return an
        instance of SerialPort.  SerialPort::open with a block passes
        a SerialPort to the block and closes it when the block exits
        (like File::open).


    ** Instance methods **

      * modem_params() -> aHash
      * modem_params=(aHash) -> aHash
      * get_modem_params() -> aHash
      * set_modem_params(aHash) -> aHash
      * set_modem_params(baudrate [, databits [, stopbits [, parity]]])

        Get and set the modem parameters.  Hash keys are "baud", "data_bits",
        "stop_bits", and "parity" (see above).

        Parameters not present in the hash or set to nil remain unchanged.
        Default parameter values for the set_modem_params method are:
        databits = 8, stopbits = 1, parity = (databits == 8 ?
        SerialPort::NONE : SerialPort::EVEN).

      * baud() -> anInteger
      * baud=(anInteger) -> anInteger
      * data_bits() -> 4, 5, 6, 7, or 8
      * data_bits=(anInteger) -> anInteger
      * stop_bits() -> 1 or 2
      * stop_bits=(anInteger) -> anInteger
      * parity() -> anInteger: SerialPort::NONE, SerialPort::EVEN,
          SerialPort::ODD, SerialPort::MARK, or SerialPort::SPACE
      * parity=(anInteger) -> anInteger

        Get and set the corresponding modem parameter.

      * flow_control() -> anInteger
      * flow_control=(anInteger) -> anInteger

        Get and set the flow control: SerialPort::NONE, SerialPort::HARD,
        SerialPort::SOFT, or (SerialPort::HARD | SerialPort::SOFT).

        Note: SerialPort::HARD mode is not supported on all platforms.
        SerialPort::HARD uses RTS/CTS handshaking; DSR/DTR is not
        supported.

      * read_timeout() -> anInteger
      * read_timeout=(anInteger) -> anInteger
      * write_timeout() -> anInteger
      * write_timeout=(anInteger) -> anInteger

        Get and set timeout values (in milliseconds) for reading and writing.
        A negative read timeout will return all the available data without
        waiting, a zero read timeout will not return until at least one
        byte is available, and a positive read timeout returns when the
        requested number of bytes is available or the interval between the
        arrival of two bytes exceeds the timeout value.

        Note: Read timeouts don't mix well with multi-threading.

        Note: Under Posix, write timeouts are not implemented.

      * break(time) -> nil

        Send a break for the given time.  

        time -> anInteger: tenths-of-a-second for the break.
        Note: Under Posix, this value is very approximate.

      * signals() -> aHash

        Return a hash with the state of each line status bit.  Keys are
        "rts", "dtr", "cts", "dsr", "dcd", and "ri".

        Note: Under Windows, the rts and dtr values are not included.

      * rts()
      * dtr()
      * cts()
      * dsr()
      * dcd()
      * ri() -> 0 or 1

      * rts=(0 or 1)
      * dtr=(0 or 1) -> 0 or 1

        Get and set the corresponding line status bit.

        Note: Under Windows, rts() and dtr() are not implemented.

-- License --

GPL

Guillaume Pierronnet <moumar@netcourrier.com>
Alan Stern <stern@rowland.harvard.edu>
Tobin Richard <tobin.richard@gmail.com>
Hector Parra <hector@hectorparra.com>
Ryan C. Payne <rpayne-oss@bullittsystems.com>


CHANGELOG
1.2.4 => 11/17/2013:    [NEW] flush_input() and flush_output() methods [Manuel A. Güílamo (MaG)]

1.2.3 => 11/04/2013:	[FIXED] Improved documentation (yard)

1.2.2 => 11/03/2013:	[FIXED] warning: already initialized constant VERSION

1.2.1 => 10/25/2013:	[FIXED] removed absurd circular dependency

1.2.0 => 10/25/2013:	[NEW] rake-compiler, bundler, and travis ci integration

1.1.0 => 05/14/2012:	[NEW] Ruby 2.0 Support [Aaron Patterson (tenderlove)] 
                        [FIXED] Support for baud rates up to 1000000 in Windows [Will Koehler (willkoehler)]

1.0.4 => 07/17/2010:	[FIXED] [Windows] No longer restricted to specific bitrates
						[FIXED] [Windows] Removed potentially insecure sprintf() use
						[FIXED] [Windows] Workaround for rb_sys_fail not checking GetLastError()
						
1.0.3 => 04/08/2010:	[FIXED] [Windows] Could not specify 10 or "COM10" or higher 
						[FIXED] [Windows] Warning passing INT instead of LONG to GetCommModemStatus

1.0.2 => 04/06/2010:	[FIXED] Passing a block into open did not properly handle a return from within the block

1.0.1 => 01/20/2010:   	[FIXED] Conditional RB_SERIAL_EXPORT needed for Visual Studio bonked GCC

1.0.0 => 01/12/2010:    [FIXED] x86_64 segmentation faults
						[NEW] Windows Ruby 1.9 support

0.7.4 => 10/12/2009:    [NEW] Conditional 1.8.6 & 1.9 support (POSIX only).
						[NEW] MinGW support.

0.7.3 => 10/09/2009:    [NEW] POSIX Ruby 1.9 support

0.7.2 => 02/09/2008:    Fix several Windows and POSIX compilation problems.
                        Make GCC compile the code without any warnings.
                        Make RDoc not be so talkative during gem installation.

0.7.1 => 31/08/2008:    Change to gemspec creation and fix posix compilation.

0.7.0 => 03/07/2008:    Major Code Cleanup

0.6.1 => 25/03/2003:    Minor changes

0.6 => 12/02/2003:      Windows support
                        Get/set modem parameters
                        Read and write timeouts
                        Open method

0.5 => 25/10/2002:      Cygwin support

0.4 => 19/09/2002:      Added more serial ports (total of 8 ports)

0.3 => 15/03/2002:      Damn, another bug found

0.2 => 14/03/2002:      A bug fixed (read() was not blocking)

0.1 => 14/03/2002:      First release


ext/native/serialport.c
/* Ruby/SerialPort
 * Guillaume Pierronnet <moumar@netcourrier.com>
 * Alan Stern <stern@rowland.harvard.edu>
 * Daniel E. Shipton <dshipton@redshiptechnologies.com>
 * Jonas Bähr <jonas.baehr@fs.ei.tum.de>
 * Ryan C. Payne <rpayne-oss@bullittsystems.com>
 *
 * This code is hereby licensed for public consumption under either the
 * GNU GPL v2 or greater.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include "serialport.h"

VALUE cSerialPort; /* serial port class */

VALUE sBaud, sDataBits, sStopBits, sParity; /* strings */
VALUE sRts, sDtr, sCts, sDsr, sDcd, sRi;

/*
 * @api private
 *
 * @see SerialPort#new
 * @see SerialPort#open
 */
static VALUE sp_create(class, _port)
   VALUE class, _port;
{
   return sp_create_impl(class, _port);
}

/*
 * Configure the serial port. You can pass a hash or multiple values
 * as separate arguments. Invalid or unsupported values will raise
 * an ArgumentError.
 *
 * When using a hash the following keys are recognized:
 * ["baud"] Integer from 50 to 256000, depending on platform
 * ["data_bits"] Integer from 5 to 8 (4 is allowed on Windows too)
 * ["stop_bits"] Integer, only allowed values are 1 or 2 (1.5 is not supported)
 * ["parity"] One of the constants NONE, EVEN or ODD (Windows allows also MARK and SPACE)
 *
 * When using separate arguments, they are interpreted as:
 *    (baud, data_bits = 8, stop_bits = 1, parity = (previous_databits == 8 ? NONE : EVEN))
 * A baudrate of nil will keep the old value.
 * The default parity depends on the number of databits configured before this function call.
 * 
 * @overload set_modem_params(baud, data_bits, stop_bits, parity)
 *    @param baud [Integer] the baud rate
 *    @param data_bits [Integer] the number of data bits
 *    @param stop_bits [Integer] the number of stop bits
 *    @param parity [Integer] the type of parity checking
 * @overload set_modem_params(hash)
 *    @param opts [Hash] the options to configure port
 *
 * @return [Hash] the original paramters
 * @raise [ArgumentError] if values are invalide or unsupported
 */
static VALUE sp_set_modem_params(argc, argv, self)
   int argc;
   VALUE *argv, self;
{
   return sp_set_modem_params_impl(argc, argv, self);
}

/*
 * Send a break for the given time
 *
 * @param time [Integer] break time in tenths-of-a-second
 * @return [nil]
 * @note (POSIX) this value is very approximate
 */
static VALUE sp_break(self, time)
   VALUE self, time;
{
   return sp_break_impl(self, time);
}

/*
 * Get the state of the DTR line 
 *
 * @note (Windows) DTR is not available
 * @return [Integer] the state of DTR line, 0 or 1
 */
static VALUE sp_get_dtr(self)
   VALUE self;
{
   return sp_get_dtr_impl(self);
}

/*
 * Get the flow control flag
 *
 * @return [Integer] the flow control flag
 * @see SerialPort#set_flow_control
 */
static VALUE sp_get_flow_control(self)
   VALUE self;
{
   return sp_get_flow_control_impl(self);
}

/*
 * Get the read timeout value
 *
 * @return [Integer] the read timeout, in milliseconds
 * @see SerialPort#set_read_timeout
 */
static VALUE sp_get_read_timeout(self)
   VALUE self;
{
   return sp_get_read_timeout_impl(self);
}

/*
 * Get the state of the RTS line
 *
 * @return [Integer] the state of RTS line, 0 or 1
 * @note (Windows) RTS is not available
 */
static VALUE sp_get_rts(self)
   VALUE self;
{
   return sp_get_rts_impl(self);
}

/*
 * Get the write timeout
 *
 * @return [Integer] the write timeout, in milliseconds
 * @note (POSIX) write timeouts are not implemented
 */
static VALUE sp_get_write_timeout(self)
   VALUE self;
{
   return sp_get_write_timeout_impl(self);
}

/*
 * Set the state of the DTR line
 *
 * @param val [Integer] the desired state of the DTR line, 0 or 1
 * @return [Integer] the original +val+ parameter
 */
static VALUE sp_set_dtr(self, val)
   VALUE self, val;
{
   return sp_set_dtr_impl(self, val);
}

/*
 * Set the flow control
 *
 * @param val [Integer] the flow control flag,
 *    +NONE+, +HARD+, +SOFT+, or (+HARD+ | +SOFT+)
 * @return [Integer] the original +val+ parameter
 * @note SerialPort::HARD mode is not supported on all platforms.
 * @note SerialPort::HARD uses RTS/CTS handshaking.
 *    DSR/DTR is not supported.
 */
static VALUE sp_set_flow_control(self, val)
   VALUE self, val;
{
   return sp_set_flow_control_impl(self, val);
}

/*
 * Set the timeout value (in milliseconds) for reading.
 * A negative read timeout will return all the available data without
 * waiting, a zero read timeout will not return until at least one
 * byte is available, and a positive read timeout returns when the
 * requested number of bytes is available or the interval between the
 * arrival of two bytes exceeds the timeout value.
 *
 * @param timeout [Integer] the read timeout in milliseconds
 * @return [Integer] the original +timeout+ parameter
 * @note Read timeouts don't mix well with multi-threading
 */
static VALUE sp_set_read_timeout(self, val)
   VALUE self, val;
{
   return sp_set_read_timeout_impl(self, val);
}

/*
 * Set the state of the RTS line
 *
 * @param val [Integer] the state of RTS line, 0 or 1
 * @return [Integer] the original +val+ parameter
 */
static VALUE sp_set_rts(self, val)
   VALUE self, val;
{
   return sp_set_rts_impl(self, val);
}

/*
 * Set a write timeout
 *
 * @param val [Integer] the write timeout in milliseconds
 * @return [Integer] the original +val+ parameter
 * @note (POSIX) write timeouts are not implemented
 */
static VALUE sp_set_write_timeout(self, val)
   VALUE self, val;
{
   return sp_set_write_timeout_impl(self, val);
}

/*
 * @private helper
 */
static void get_modem_params(self, mp)
   VALUE self;
   struct modem_params *mp;
{
   get_modem_params_impl(self, mp);
}

/*
 * Set the baud rate
 *
 * @param data_rate [Integer] the baud rate
 * @return [Integer] the original +data_rate+ parameter
 * @see SerialPort#set_modem_params
 */
static VALUE sp_set_data_rate(self, data_rate)
   VALUE self, data_rate;
{
   VALUE argv[4];

   argv[0] = data_rate;
   argv[1] = argv[2] = argv[3] = Qnil;
   sp_set_modem_params(4, argv, self);

   return data_rate;
}

/*
 * Set the data bits
 *
 * @param data_bits [Integer] the number of data bits
 * @return [Integer] the original +data_bits+ parameter
 * @see SerialPort#set_modem_params
 */
static VALUE sp_set_data_bits(self, data_bits)
   VALUE self, data_bits;
{
   VALUE argv[4];

   argv[1] = data_bits;
   argv[0] = argv[2] = argv[3] = Qnil;
   sp_set_modem_params(4, argv, self);

   return data_bits;
}

/*
 * Set the stop bits
 *
 * @param stop_bits [Integer] the number of stop bits
 * @return [Integer] the original +stop_bits+ parameter
 * @see SerialPort#set_modem_params
 */
static VALUE sp_set_stop_bits(self, stop_bits)
   VALUE self, stop_bits;
{
   VALUE argv[4];

   argv[2] = stop_bits;
   argv[0] = argv[1] = argv[3] = Qnil;
   sp_set_modem_params(4, argv, self);

   return stop_bits;
}

/*
 * Set the parity
 *
 * @param parity [Integer] the parity type
 * @return [Integer] the original +parity+ parameter
 * @see SerialPort#set_modem_params
 */
static VALUE sp_set_parity(self, parity)
   VALUE self, parity;
{
   VALUE argv[4];

   argv[3] = parity;
   argv[0] = argv[1] = argv[2] = Qnil;
   sp_set_modem_params(4, argv, self);

   return parity;
}

/*
 * Get the current baud rate
 *
 * @return [Integer] the current baud rate
 * @see SerialPort#set_modem_params
 */
static VALUE sp_get_data_rate(self)
   VALUE self;
{
   struct modem_params mp;

   get_modem_params(self, &mp);

   return INT2FIX(mp.data_rate);
}

/*
 * Get the current data bits
 *
 * @return [Integer] the current number of data bits
 * @see SerialPort#set_modem_params
 */
static VALUE sp_get_data_bits(self)
   VALUE self;
{
   struct modem_params mp;

   get_modem_params(self, &mp);

   return INT2FIX(mp.data_bits);
}

/*
 * Get the current stop bits
 *
 * @return [Integer] the current number of stop bits
 * @see SerialPort#set_modem_params for details
 */
static VALUE sp_get_stop_bits(self)
   VALUE self;
{
   struct modem_params mp;

   get_modem_params(self, &mp);

   return INT2FIX(mp.stop_bits);
}

/*
 * Get the current parity
 *
 * @return [Integer] the current parity
 * @see SerialPort#set_modem_params
 */
static VALUE sp_get_parity(self)
   VALUE self;
{
   struct modem_params mp;

   get_modem_params(self, &mp);

   return INT2FIX(mp.parity);
}

/*
 * Get the configure of the serial port
 *
 * @return [Hash] the serial port configuration
 * @see SerialPort#set_modem_params
 */
static VALUE sp_get_modem_params(self)
   VALUE self;
{
   struct modem_params mp;
   VALUE hash;

   get_modem_params(self, &mp);

   hash = rb_hash_new();

   rb_hash_aset(hash, sBaud, INT2FIX(mp.data_rate));
   rb_hash_aset(hash, sDataBits, INT2FIX(mp.data_bits));
   rb_hash_aset(hash, sStopBits, INT2FIX(mp.stop_bits));
   rb_hash_aset(hash, sParity, INT2FIX(mp.parity));

   return hash;
}

/*
 * @api private
 */
void get_line_signals_helper(obj, ls)
   VALUE obj;
   struct line_signals *ls;
{
   get_line_signals_helper_impl(obj, ls);
}

/*
 * Get the state of the CTS line
 *
 * @return [Integer] the state of the CTS line, 0 or 1
 * @see SerialPort#get_signals
 */
static VALUE sp_get_cts(self)
   VALUE self;
{
   struct line_signals ls;

   get_line_signals_helper(self, &ls);

   return INT2FIX(ls.cts);
}

/*
 * Get the state of the DSR line
 *
 * @return [Integer] the state of the DSR line, 0 or 1
 * @see SerialPort#get_signals
 */
static VALUE sp_get_dsr(self)
   VALUE self;
{
   struct line_signals ls;

   get_line_signals_helper(self, &ls);

   return INT2FIX(ls.dsr);
}

/*
 * Get the state of the DCD line
 *
 * @return [Integer] the state of the DCD line, 0 or 1
 * @see SerialPort#get_signals
 */
static VALUE sp_get_dcd(self)
   VALUE self;
{
   struct line_signals ls;

   get_line_signals_helper(self, &ls);

   return INT2FIX(ls.dcd);
}

/*
 * Get the state of the RI line
 *
 * @return [Integer] the state of the RI line, 0 or 1
 * @see SerialPort#get_signals
 */
static VALUE sp_get_ri(self)
   VALUE self;
{
   struct line_signals ls;

   get_line_signals_helper(self, &ls);

   return INT2FIX(ls.ri);
}

/*
 * Return a hash with the state of each line status bit.
 * Keys:
 *    "rts", "dtr", "cts", "dsr", "dcd", and "ri".
 *
 * @return [Hash] the state line info
 * @note (Windows) the rts and dtr values are not included
 * @note This method is implemented as both SerialPort#signals and SerialPort#get_signals
 */
static VALUE sp_signals(self)
   VALUE self;
{
   struct line_signals ls;
   VALUE hash;

   get_line_signals_helper(self, &ls);

   hash = rb_hash_new();

#if !(defined(OS_MSWIN) || defined(OS_BCCWIN) || defined(OS_MINGW))
   rb_hash_aset(hash, sRts, INT2FIX(ls.rts));
   rb_hash_aset(hash, sDtr, INT2FIX(ls.dtr));
#endif
   rb_hash_aset(hash, sCts, INT2FIX(ls.cts));
   rb_hash_aset(hash, sDsr, INT2FIX(ls.dsr));
   rb_hash_aset(hash, sDcd, INT2FIX(ls.dcd));
   rb_hash_aset(hash, sRi, INT2FIX(ls.ri));

   return hash;
}

/**
 * Flush data received but not read.
 *
 * @return [Boolean] true on success otherwise an error occurs. It causes all
 * errors the tcflush(3) system call causes. The caller should care such
 * errors.
 * @note (Windows) flush_input is not available
 */
static VALUE sp_flush_input_data(self)
   VALUE self;
{
   return sp_flush_input_data_impl(self);
}

/**
 * Flush data written but not transmitted.
 *
 * @return [Boolean] true on success otherwise an error occurs. It causes all
 * errors the tcflush(3) system call causes. The caller should care such
 * errors.
 * @note (Windows) flush_output is not available
 */
static VALUE sp_flush_output_data(self)
   VALUE self;
{
   return sp_flush_output_data_impl(self);
}

/**
 * Flush both data received but not read and data written but not transmitted.
 *
 * @return [Boolean] true on success otherwise an error occurs. It causes all
 * errors the tcflush(3) system call causes. The caller should care such
 * errors.
 * @note (Windows) flush is not available
 */
static VALUE sp_flush_all_data(self)
   VALUE self;
{
   return sp_flush_all_data_impl(self);
}


void Init_serialport()
{
   sBaud = rb_str_new2("baud");
   sDataBits = rb_str_new2("data_bits");
   sStopBits = rb_str_new2("stop_bits");
   sParity = rb_str_new2("parity");
   sRts = rb_str_new2("rts");
   sDtr = rb_str_new2("dtr");
   sCts = rb_str_new2("cts");
   sDsr = rb_str_new2("dsr");
   sDcd = rb_str_new2("dcd");
   sRi = rb_str_new2("ri");

   rb_gc_register_address(&sBaud);
   rb_gc_register_address(&sDataBits);
   rb_gc_register_address(&sStopBits);
   rb_gc_register_address(&sParity);
   rb_gc_register_address(&sRts);
   rb_gc_register_address(&sDtr);
   rb_gc_register_address(&sCts);
   rb_gc_register_address(&sDsr);
   rb_gc_register_address(&sDcd);
   rb_gc_register_address(&sRi);

   cSerialPort = rb_define_class("SerialPort", rb_cIO);
   rb_define_singleton_method(cSerialPort, "create", sp_create, 1);

   rb_define_method(cSerialPort, "get_modem_params", sp_get_modem_params, 0);
   rb_define_method(cSerialPort, "set_modem_params", sp_set_modem_params, -1);
   rb_define_method(cSerialPort, "modem_params", sp_get_modem_params, 0);
   rb_define_method(cSerialPort, "modem_params=", sp_set_modem_params, -1);
   rb_define_method(cSerialPort, "baud", sp_get_data_rate, 0);
   rb_define_method(cSerialPort, "baud=", sp_set_data_rate, 1);
   rb_define_method(cSerialPort, "data_bits", sp_get_data_bits, 0);
   rb_define_method(cSerialPort, "data_bits=", sp_set_data_bits, 1);
   rb_define_method(cSerialPort, "stop_bits", sp_get_stop_bits, 0);
   rb_define_method(cSerialPort, "stop_bits=", sp_set_stop_bits, 1);
   rb_define_method(cSerialPort, "parity", sp_get_parity, 0);
   rb_define_method(cSerialPort, "parity=", sp_set_parity, 1);

   rb_define_method(cSerialPort, "flow_control=", sp_set_flow_control, 1);
   rb_define_method(cSerialPort, "flow_control", sp_get_flow_control, 0);

   rb_define_method(cSerialPort, "read_timeout", sp_get_read_timeout, 0);
   rb_define_method(cSerialPort, "read_timeout=", sp_set_read_timeout, 1);
   rb_define_method(cSerialPort, "write_timeout", sp_get_write_timeout, 0);
   rb_define_method(cSerialPort, "write_timeout=", sp_set_write_timeout, 1);

   rb_define_method(cSerialPort, "break", sp_break, 1);

   rb_define_method(cSerialPort, "signals", sp_signals, 0);
   rb_define_method(cSerialPort, "get_signals", sp_signals, 0);
   rb_define_method(cSerialPort, "rts", sp_get_rts, 0);
   rb_define_method(cSerialPort, "rts=", sp_set_rts, 1);
   rb_define_method(cSerialPort, "dtr", sp_get_dtr, 0);
   rb_define_method(cSerialPort, "dtr=", sp_set_dtr, 1);
   rb_define_method(cSerialPort, "cts", sp_get_cts, 0);
   rb_define_method(cSerialPort, "dsr", sp_get_dsr, 0);
   rb_define_method(cSerialPort, "dcd", sp_get_dcd, 0);
   rb_define_method(cSerialPort, "ri", sp_get_ri, 0);

   rb_define_method(cSerialPort, "flush_input", sp_flush_input_data, 0);
   rb_define_method(cSerialPort, "flush_output", sp_flush_output_data, 0);
   rb_define_method(cSerialPort, "flush", sp_flush_all_data, 0);

   /*
    * 0
    */
   rb_define_const(cSerialPort, "NONE", INT2FIX(NONE));

   /*
    * 1
    */
   rb_define_const(cSerialPort, "HARD", INT2FIX(HARD));

   /*
    * 2
    */
   rb_define_const(cSerialPort, "SOFT", INT2FIX(SOFT));

   /*
    * 0
    */
   rb_define_const(cSerialPort, "SPACE", INT2FIX(SPACE));

   /*
    * 1
    */
   rb_define_const(cSerialPort, "MARK", INT2FIX(MARK));

   /*
    * 2
    */
   rb_define_const(cSerialPort, "EVEN", INT2FIX(EVEN));

   /*
    * 3
    */
   rb_define_const(cSerialPort, "ODD", INT2FIX(ODD));
}
